### PR TITLE
add dockerized tslint

### PR DESCRIPTION
**Pull request by BenTheElder**
https://github.com/kubernetes/test-infra/issues/23796, https://github.com/kubernetes/test-infra/issues/23829

### CODE REVIEW

**Comment by BenTheElder** [comment]
/hold
I forgot one thing 🙃

File: hack/run-in-node-container.sh
Comment (BenTheElder) [inline hack/run-in-node-container.sh:30]: yarn tries to read $HOME/.yarnrc and put a cache under $HOME/.cache, when we change the user in the line above (so output is correctly owned on the host) the user doesn't exist and yarn fails on inability to open `/.yarnrc`. Setting it to a valid writeable location allows yarn to function and is pretty harmless, yarn will be writing a system cache copy of package on install under `/tmp/.cache`.

The rest of this is hopefully more self-explanatory.

File: hack/make-rules/verify/tslint.sh
Comment (BenTheElder) [inline hack/make-rules/verify/tslint.sh:31]: this takes maybe 2s (some variability), the whole `tslint.sh` takes ~6s total. not great, but hardly the slowest thing we're doing. 

probably not worth the complexity trying to sometimes skip this?
we already have the yarn lock and `node_modules/` cache

**Comment by BenTheElder** [comment]
If I introduce a trivial error into one of the typescript files this definitely catches it, but I'm not sure yet that it has all the same options as under bazel.

File: .node-image
Comment (BenTheElder) [inline .node-image:1]: I'm not sure adding all these one-off version files is great. In this case it might be fine to just stick it in `run-in-node-container.sh` since we should only need node at build-time and we can do that all with that script AFAICT.

Originally I thought we might need to do some caching, but gitignored `node_modules` seems to handle that fine.

regarding the choice of image:
- relatively small, but complete
- we are on node 12.X under bazel currently

**Comment by BenTheElder** [comment]
This is currently failing because `pull-test-infra-verify-lint` doesn't have docker running.

We should also consider investigating something like https://github.com/nvm-sh/nvm instead, similar to how we're now using gimmee for go, but containerizing node is very straightforward and avoids interacting with host state (other than the repo dir) which I know we can do fairly easily with go binaries on-the-host-but-not-system-installed but less sure with the node ecosystem.

File: hack/make-rules/verify/tslint.sh
Comment (chaodaiG) [reply hack/make-rules/verify/tslint.sh:31]: 2s should be fine, we do have many other stuffs that take much longer

File: hack/run-in-node-container.sh
Comment (chaodaiG) [reply hack/run-in-node-container.sh:30]: should this explanation be a comment above `--user` line?

File: .node-image
Comment (chaodaiG) [reply .node-image:1]: I like the idea of sticking in in `run-in-node-container.sh`, compilation for `deck` can also be done in a way similar to `yarn install`, we should be fine

**Review by chaodaiG — COMMENTED** [review]
Looks good to me! Also add a target after https://github.com/kubernetes/test-infra/blob/724022ee191844c1a8ebc0a1f6505ddb3694844e/Makefile#L85?

**Comment by chaodaiG** [comment]
> We should also consider investigating something like https://github.com/nvm-sh/nvm instead, similar to how we're now using gimmee for go, but containerizing node is very straightforward and avoids interacting with host state (other than the repo dir) which I know we can do fairly easily with go binaries on-the-host-but-not-system-installed but less sure with the node ecosystem.

At this point, I think it makes more sense to go with docker approach to ensure hermetic build/test. We can iterate on nvm once there is a domain expert become interested in this

File: .node-image
Comment (BenTheElder) [reply .node-image:1]: I'll do this.

File: hack/make-rules/verify/tslint.sh
Comment (BenTheElder) [reply hack/make-rules/verify/tslint.sh:31]: it will be something like this *per* typescript target though, unless we split it out and use a make dependency or something.

right now it will avoid re-downloading everything, but 2s is a bit slower than I'd like to do at the beginning of each typescript related script/target.

File: hack/run-in-node-container.sh
Comment (BenTheElder) [reply hack/run-in-node-container.sh:30]: I can do it, but we'll have to abuse a bash trick ;-)

technically normal comments don't work between line continuation (`\`)

Comment (chaodaiG) [reply hack/run-in-node-container.sh:30]: Add it above `docker run`?

File: hack/make-rules/verify/tslint.sh
Comment (chaodaiG) [reply hack/make-rules/verify/tslint.sh:31]: I think it depends on whether anyone would do this locally or not, maybe `make verify-tslint`, doubt anyone would need to compile deck locally very frequently.

For prow jobs, I think `lint` job does it once, `integration-test` compiles deck once, and `push-prow-image` also compiles deck once, each with 2s overhead.

so we might be fine?

Comment (BenTheElder) [reply hack/make-rules/verify/tslint.sh:31]: it's also something we can improve on later, factoring it out into a seperate script will make that relatively easy to make a make target dependency later

File: .node-image
Comment (BenTheElder) [reply .node-image]: done

File: hack/run-in-node-container.sh
Comment (BenTheElder) [reply hack/run-in-node-container.sh:38]: I wound up doing that :-)

You can do this for future reference:
```bash
foo \
 `#bar comment` \
  baz
```

This works with backticks (which some shells considered deprecated vs `$()` but realistically cannot ever remove) and does not actually create a subshell. But when I discovered it trying to find a good way to comment a more complex docker invocation in POSIX shell I couldn't find many examples of it in the wild.

In bash the decent answer is to place the args in an array and comment the array instead.

```bash
docker_options = (
    # bar comment
    'foo'
    # quux comment
    'baz'
)
docker run "${docker_options[@]}"
```

**Comment by chaodaiG** [comment]
enabling dind in https://github.com/kubernetes/test-infra/pull/24540, should unblock this PR

**Comment by chaodaiG** [comment]
```
hack/build/ensure-node_modules.sh: line 26: hack/run-in-node-container.sh: No such file or directory
```
chmod?

**Comment by BenTheElder** [comment]
no, but one too many `/..` in the `REPO_ROOT` snippet

**Review by chaodaiG — APPROVED** [review]

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/issues/24200
/retest

**Pull request merged**
